File: elixir/lib/bonemesh/handshake.ex
defmodule Bonemesh.Handshake do
  @moduledoc """
  The BMX handshake (security.md §4): a three-message, mutually authenticated,
  forward-secret exchange. Hybrid X25519 + ML-KEM-768 forward secrecy mixed
  through `Bonemesh.KeySchedule`; authentication by a root-signed certificate
  plus an ML-DSA signature over the live transcript. Field-for-field identical
  to the Java reference, so an Elixir node and a Java node complete a handshake
  together.

  A handshake value is a struct threaded through the step functions; each step
  returns `{message, state}` or `{:ok, message, state}` / `{:error, reason}`.
  """

  alias Bonemesh.{Cert, Crypto, KeySchedule}

  @version 3

  defstruct [
    :initiator,
    :mesh,
    :root_public,
    :now,
    :cert,
    :id_public,
    :id_private,
    :ks,
    :eph_dh_public,
    :eph_dh_private,
    :eph_kem_public,
    :eph_kem_private,
    :session,
    :peer_cert
  ]

  @doc "Creates the initiating side."
  def initiator(mesh, root_public, now, cert, id_public, id_private),
    do: new(true, mesh, root_public, now, cert, id_public, id_private)

  @doc "Creates the responding side."
  def responder(mesh, root_public, now, cert, id_public, id_private),
    do: new(false, mesh, root_public, now, cert, id_public, id_private)

  defp new(initiator, mesh, root_public, now, cert, id_public, id_private) do
    ks = KeySchedule.new() |> KeySchedule.mix_hash(mesh)

    %__MODULE__{
      initiator: initiator,
      mesh: mesh,
      root_public: root_public,
      now: now,
      cert: cert,
      id_public: id_public,
      id_private: id_private,
      ks: ks
    }
  end

  @doc "Initiator: produces message 1."
  def write_message1(%__MODULE__{initiator: true} = s) do
    {dh_pub, dh_priv} = Crypto.x25519_generate()
    {kem_pub, kem_priv} = Crypto.mlkem_generate()
    n = :crypto.strong_rand_bytes(32)

    ks = s.ks |> KeySchedule.mix_hash(dh_pub) |> KeySchedule.mix_hash(kem_pub) |> KeySchedule.mix_hash(n)

    msg =
      line(%{
        "t" => "bmx1",
        "v" => @version,
        "mesh" => s.mesh,
        "e" => Base.encode64(dh_pub),
        "k" => Base.encode64(kem_pub),
        "n" => Base.encode64(n)
      })

    {msg, %{s | eph_dh_public: dh_pub, eph_dh_private: dh_priv, eph_kem_public: kem_pub, eph_kem_private: kem_priv, ks: ks}}
  end

  @doc "Responder: consumes message 1, produces message 2."
  def read_message1_write_message2(%__MODULE__{initiator: false} = s, msg1) do
    # Malformed peer input (bad JSON, bad base64, wrong shape) is rejected
    # gracefully rather than crashing the connection process.
    try do
      do_read_message1_write_message2(s, msg1)
    rescue
      _ -> {:error, "malformed handshake message"}
    end
  end

  defp do_read_message1_write_message2(s, msg1) do
    m = decode(msg1)

    cond do
      m["t"] != "bmx1" -> {:error, "expected bmx1"}
      m["v"] != @version -> {:error, "unsupported version"}
      m["mesh"] != s.mesh -> {:error, "mesh mismatch"}
      true ->
        ei_pub = Base.decode64!(m["e"])
        ki_ek = Base.decode64!(m["k"])
        n = Base.decode64!(m["n"])

        ks = s.ks |> KeySchedule.mix_hash(ei_pub) |> KeySchedule.mix_hash(ki_ek) |> KeySchedule.mix_hash(n)

        {er_pub, er_priv} = Crypto.x25519_generate()
        ks = KeySchedule.mix_hash(ks, er_pub)
        ss_dh = Crypto.x25519_agree(ei_pub, er_priv)
        ks = KeySchedule.mix_key(ks, ss_dh)

        {ss_kem, ct} = Crypto.mlkem_encapsulate(ki_ek)
        ks = ks |> KeySchedule.mix_hash(ct) |> KeySchedule.mix_key(ss_kem)

        {auth, ks} = seal_identity(ks, s)

        msg =
          line(%{"t" => "bmx2", "e" => Base.encode64(er_pub), "ct" => Base.encode64(ct), "auth" => Base.encode64(auth)})

        {:ok, msg, %{s | ks: ks}}
    end
  end

  @doc "Initiator: consumes message 2 (verifying the responder), produces message 3."
  def read_message2_write_message3(%__MODULE__{initiator: true} = s, msg2) do
    try do
      do_read_message2_write_message3(s, msg2)
    rescue
      _ -> {:error, "malformed handshake message"}
    end
  end

  defp do_read_message2_write_message3(s, msg2) do
    m = decode(msg2)
    er_pub = Base.decode64!(m["e"])
    ct = Base.decode64!(m["ct"])
    auth = Base.decode64!(m["auth"])

    ks = KeySchedule.mix_hash(s.ks, er_pub)
    ss_dh = Crypto.x25519_agree(er_pub, s.eph_dh_private)
    ks = KeySchedule.mix_key(ks, ss_dh)
    ks = KeySchedule.mix_hash(ks, ct)
    ss_kem = Crypto.mlkem_decapsulate(s.eph_kem_private, ct)
    ks = KeySchedule.mix_key(ks, ss_kem)

    case open_identity(ks, auth, s) do
      {:error, reason} ->
        {:error, reason}

      {:ok, peer_cert, ks} ->
        {auth_i, ks} = seal_identity(ks, s)
        msg = line(%{"t" => "bmx3", "auth" => Base.encode64(auth_i)})
        {i2r, r2i} = KeySchedule.split(ks)
        session = %{send_key: i2r, receive_key: r2i, peer_cert: peer_cert}
        {:ok, msg, %{s | ks: ks, session: session, peer_cert: peer_cert}}
    end
  end

  @doc "Responder: consumes message 3, completing the handshake."
  def read_message3(%__MODULE__{initiator: false} = s, msg3) do
    try do
      do_read_message3(s, msg3)
    rescue
      _ -> {:error, "malformed handshake message"}
    end
  end

  defp do_read_message3(s, msg3) do
    m = decode(msg3)
    auth = Base.decode64!(m["auth"])

    case open_identity(s.ks, auth, s) do
      {:error, reason} ->
        {:error, reason}

      {:ok, peer_cert, ks} ->
        {i2r, r2i} = KeySchedule.split(ks)
        session = %{send_key: r2i, receive_key: i2r, peer_cert: peer_cert}
        {:ok, %{s | ks: ks, session: session, peer_cert: peer_cert}}
    end
  end

  @doc "The completed session (`%{send_key, receive_key, peer_cert}`)."
  def session(%__MODULE__{session: session}), do: session

  # Signs the current transcript, packages {cert, sig}, encrypts into the transcript.
  defp seal_identity(ks, s) do
    sig = Crypto.mldsa_sign(:mldsa65, s.id_private, ks.h)
    payload = JSON.encode!(%{"cert" => s.cert, "sig" => Base.encode64(sig)})
    KeySchedule.encrypt_and_hash(ks, payload)
  end

  # Decrypts the peer identity, verifies its certificate and transcript signature.
  defp open_identity(ks, auth, s) do
    h_pre = ks.h

    case KeySchedule.decrypt_and_hash(ks, auth) do
      :error ->
        {:error, "handshake authentication failed"}

      {:ok, plaintext, ks} ->
        payload = JSON.decode!(plaintext)
        peer_cert = payload["cert"]
        sig = Base.decode64!(payload["sig"])

        with :ok <- Cert.verify(peer_cert, s.root_public, s.mesh, s.now),
             true <- Crypto.mldsa_verify(:mldsa65, Cert.identity_key(peer_cert), h_pre, sig) do
          {:ok, peer_cert, ks}
        else
          {:error, reason} -> {:error, "peer certificate invalid: " <> reason}
          false -> {:error, "peer transcript signature does not verify"}
        end
    end
  end

  defp line(map), do: JSON.encode!(map) <> "\n"
  defp decode(bin), do: bin |> String.trim() |> JSON.decode!()
end

File: python/bonemesh/handshake.py
"""BMX handshake (security.md §4).

A three-message, mutually authenticated, forward-secret exchange: hybrid
X25519 + ML-KEM-768 forward secrecy through the key schedule, authentication by a
root-signed certificate plus an ML-DSA signature over the live transcript.
Field-for-field identical to every other implementation.

Two places where the shipped wire differs from a literal reading of the spec
prose, and where this port follows the code the other six actually run:

* ``bmx2``/``bmx3`` carry a single ``auth`` field -- the sealed
  ``{"cert": ..., "sig": ...}`` payload -- rather than separate ``cert`` and
  ``sig`` members.
* The transcript absorbs individual field *values* in order, not the raw wire
  bytes of each message.

The ``write_*`` methods return an encoded frame ready for the wire; the ``read_*``
methods take the peer's already-decoded frame dict.
"""

from __future__ import annotations

import base64
import json
import secrets

from bonemesh import cert as certmod
from bonemesh import crypto as c
from bonemesh.frame import encode
from bonemesh.keyschedule import KeySchedule


def _b64(b: bytes) -> str:
    return base64.b64encode(b).decode("ascii")


def _unb64(s: str) -> bytes:
    return base64.b64decode(s)


class HandshakeError(Exception):
    pass


class Session:
    """The completed handshake's outputs."""

    __slots__ = ("send_key", "receive_key", "peer_cert", "h")

    def __init__(self, send_key: bytes, receive_key: bytes, peer_cert: dict, h: bytes) -> None:
        self.send_key = send_key
        self.receive_key = receive_key
        self.peer_cert = peer_cert
        self.h = h


class Handshake:
    def __init__(self, mesh: str, root_public: bytes, now: int, cert_obj: dict,
                 id_private: bytes) -> None:
        self.mesh = mesh
        self.root_public = root_public
        self.now = now
        self.cert = cert_obj
        self.id_private = id_private
        self.ks = KeySchedule()
        self.ks.mix_hash(mesh.encode("utf-8"))
        self.session: Session | None = None
        self._eph_dh_priv = None
        self._eph_kem_dk = None

    @classmethod
    def initiator(cls, mesh, root_public, now, cert_obj, id_private) -> "Handshake":
        return cls(mesh, root_public, now, cert_obj, id_private)

    @classmethod
    def responder(cls, mesh, root_public, now, cert_obj, id_private) -> "Handshake":
        return cls(mesh, root_public, now, cert_obj, id_private)

    def write_message1(self) -> bytes:
        """Message 1 (initiator)."""
        dh_pub, dh_priv = c.x25519_generate()
        kem_ek, kem_dk = c.mlkem768_keypair()
        n = secrets.token_bytes(32)
        self._eph_dh_priv = dh_priv
        self._eph_kem_dk = kem_dk
        self.ks.mix_hash(dh_pub)
        self.ks.mix_hash(kem_ek)
        self.ks.mix_hash(n)
        return encode({"t": "bmx1", "v": 3, "mesh": self.mesh,
                       "e": _b64(dh_pub), "k": _b64(kem_ek), "n": _b64(n)})

    def read_message1_write_message2(self, m: dict) -> bytes:
        """Message 2 (responder): consume msg1, produce msg2."""
        if m.get("t") != "bmx1":
            raise HandshakeError("expected bmx1")
        if m.get("v") != 3:
            raise HandshakeError("unsupported version")
        if m.get("mesh") != self.mesh:
            raise HandshakeError("mesh mismatch")
        self.ks.mix_hash(_unb64(m["e"]))
        self.ks.mix_hash(_unb64(m["k"]))
        self.ks.mix_hash(_unb64(m["n"]))

        er_pub, er_priv = c.x25519_generate()
        self.ks.mix_hash(er_pub)
        self.ks.mix_key(c.x25519_agree(er_priv, _unb64(m["e"])))

        ss, ct = c.mlkem768_encapsulate(_unb64(m["k"]))
        self.ks.mix_hash(ct)
        self.ks.mix_key(ss)

        auth = self._seal_identity()
        return encode({"t": "bmx2", "e": _b64(er_pub), "ct": _b64(ct), "auth": _b64(auth)})

    def read_message2_write_message3(self, m: dict) -> bytes:
        """Message 3 (initiator): verify the responder, produce msg3."""
        er_pub = _unb64(m["e"])
        ct = _unb64(m["ct"])
        auth = _unb64(m["auth"])

        self.ks.mix_hash(er_pub)
        self.ks.mix_key(c.x25519_agree(self._eph_dh_priv, er_pub))
        self.ks.mix_hash(ct)
        ss_kem = c.mlkem768_decapsulate(self._eph_kem_dk, ct)
        if ss_kem is None:
            raise HandshakeError("decapsulation failed")
        self.ks.mix_key(ss_kem)

        peer_cert = self._open_identity(auth)
        auth_i = self._seal_identity()
        out = encode({"t": "bmx3", "auth": _b64(auth_i)})
        i2r, r2i = self.ks.split()
        self.session = Session(i2r, r2i, peer_cert, self.ks.h)
        return out

    def read_message3(self, m: dict) -> None:
        """Message 3 (responder): verify the initiator, completing the handshake."""
        peer_cert = self._open_identity(_unb64(m["auth"]))
        i2r, r2i = self.ks.split()
        self.session = Session(r2i, i2r, peer_cert, self.ks.h)

    def _seal_identity(self) -> bytes:
        sig = c.mldsa65_sign(self.id_private, self.ks.h)
        payload = json.dumps({"cert": self.cert, "sig": _b64(sig)},
                             separators=(",", ":"), ensure_ascii=False).encode("utf-8")
        return self.ks.encrypt_and_hash(payload)

    def _open_identity(self, auth: bytes) -> dict:
        # The signature pre-image is h as it stands immediately BEFORE the
        # decrypt absorbs the ciphertext, so it has to be captured first.
        h_pre = self.ks.h
        pt = self.ks.decrypt_and_hash(auth)
        if pt is None:
            raise HandshakeError("handshake authentication failed")
        try:
            payload = json.loads(pt.decode("utf-8"))
        except (ValueError, UnicodeDecodeError) as e:
            raise HandshakeError("bad auth payload") from e
        peer_cert = payload.get("cert")
        if not isinstance(peer_cert, dict):
            raise HandshakeError("bad auth payload")
        reason = certmod.verify(peer_cert, self.root_public, self.mesh, self.now)
        if reason:
            raise HandshakeError(f"peer certificate invalid: {reason}")
        idk = certmod.identity_key(peer_cert)
        if not c.mldsa65_verify(idk, h_pre, _unb64(payload["sig"])):
            raise HandshakeError("peer transcript signature does not verify")
        return peer_cert
